### PR TITLE
[Quick Add Bulk] Adjust scrolling the Quick Order List to work inside modal

### LINKED ISSUE #3312: [Quick Add bulk] Quick Order List scrolling adjustment for Keyboard interaction
The Keyboard interaction scrolling in Quick order List needs to be adjusted for the case where we show the Quick Order List inside the Quick Add Bulk modal. 

Currently the scrolling doesn't happen when you reach a not fully shown row/input.

29-59-5nk6c-xyyg4

### PR DESCRIPTION
**Pull request by eugenekasimov**
### PR Summary: 

This PR makes sure that scrolling of the Quick Order List inside the modal of the Quick Add Bulk works properly. 

### Why are these changes introduced?

Fixes #3312.

[Befor](https://screenshot.click/13-57-pvzz6-3v6l6.mp4)

[After](https://screenshot.click/13-01-goa06-b1ez4.mp4)

### What approach did you take?

-  I added a condition depending whether the Quick Order List is rendered inside the Quick Add Bulk modal or not. If yes, then we calculate positions for table head and total bar. 
- I also moved all logic related to scrolling to a new function`scrollQuickOrderListTable() ` I believe it makes sense and the code looks more readable.

### Other considerations
In the feature we would want to trigger scrolling effect on TABing buttons. This PR covers only a case when you move over the list via inputs by hitting Enter. 

### Testing steps/scenarios

- [ ] Go to Shop All collection. In the Product grid template make sure Quick Add set to Bulk. 
- [ ] On any product cards click on Choose options to open a modal with the Quick Order List inside. 
- [ ] Start moving down/up the inputs by hitting Enter/Shift+Enter. Make sure that as soon as you reach an input that is partially hidden under the total bar or table header the list is automatically scrolled. 
- [ ] Make sure that the same pattern in the Quick Order List on the PDP hasn't been affected and works the same way. 

### Demo links

- [Editor](https://admin.shopify.com/store/os2-demo/themes/163020111894/editor?previewPath=%2Fcollections%2Fall&section=template--21573248581654__product-grid)

### Checklist
- [ ] Added PR summary for [release notes](https://themes.shopify.com/themes/dawn/styles/default#ReleaseNotes)
- [ ] Requested review from UX (Only for changes that are affecting the experience or perceivable visual details)
- [ ] Created a ticket for the [help.shopify.com](https://help.shopify.com) documentation team about updates to theme settings. (Internal-only task)
- [ ] Followed [theme code principles](https://github.com/Shopify/dawn/blob/main/.github/CONTRIBUTING.md#theme-code-principles)
- [ ] Linted with [Theme Check](https://github.com/Shopify/theme-check)
- [ ] Tested on [mobile](https://shopify.dev/themes/store/requirements#mobile-browser-requirements)
- [ ] Tested on [multiple browsers](https://shopify.dev/themes/store/requirements#desktop-browser-requirements)
- [ ] Tested for [accessibility](https://shopify.dev/themes/best-practices/accessibility)

### CODE REVIEW

File: assets/quick-add.css [hunk 1]
```diff
@@ -88,7 +88,7 @@
 
   .quick-add-modal__content-info--bulk thead {
     position: sticky;
-    z-index: 2;
+    z-index: 3;
```
Comment (utkarshsaxena93) [inline assets/quick-add.css:91]: Why this change? Should we make this a sufficiently large value if we want this to appear on the top at all times?

**Review by utkarshsaxena93 — APPROVED** [review]
Scroll behaviour looks good.

Additional improvement to how we manage focus https://share.descript.com/view/MZeAfZs5K8s

This should be the default behaviour for all modals where it traps focus, not just this one.

**Review by melissaperreault — COMMENTED** [review]
Aligned with @utkarshsaxena93 's comments! We do have the close X sticky on smaller screens, might be quick to add on larger screens!? 🤷 Agree it's a follow up!

**_Android and iPhone_**
- [ ] The input in the modal keeps selecting the value and reveals other UI. It's not the same behaviour from other input component elsewhere. It is annoying as a user. 

| Modal | Product template | Cart |
|--------|--------|--------|
| Screenshot_20240314-120539_Chrome | Screenshot_20240314-120555_Chrome | Screenshot_20240314-120609_Chrome |

**Comment by eugenekasimov** [comment]
> Scroll behaviour looks good.
> 
> Additional improvement to how we manage focus https://share.descript.com/view/MZeAfZs5K8s
> 
> This should be the default behaviour for all modals where it traps focus, not just this one.

Thanks for the feedback Utkarsh. About the taping focus I absolutely agree with you. Since it's not related to this PR I'll create a ticket for that and also I'll check if we have this issue on other modals in the theme 👍

The comment about close button "X" I believe @sofiamatulis may have context on it because I noticed it in our base PR Quick Add Bulk and I think I even mentioned about it but she said it was changed that way on purpose.

**Comment by sofiamatulis** [comment]
> > Scroll behaviour looks good.
> > Additional improvement to how we manage focus https://share.descript.com/view/MZeAfZs5K8s
> > This should be the default behaviour for all modals where it traps focus, not just this one.
> 
> Thanks for the feedback Utkarsh. About the taping focus I absolutely agree with you. Since it's not related to this PR I'll create a ticket for that and also I'll check if we have this issue on other modals in the theme 👍
> 
> The comment about close button "X" I believe @sofiamatulis may have context on it because I noticed it in our base PR Quick Add Bulk and I think I even mentioned about it but she said it was changed that way on purpose.

Yes, the x button being sticky only on mobile is a behavior we have for the existing modal. But we can challenge that experience and create a follow up ticket :) (although unrelated to this project since it also affect the existing modal)

File: assets/quick-add.css (hunk 1 above)
Comment (eugenekasimov) [reply assets/quick-add.css:91]: Here is why https://share.descript.com/view/wOls6SoXhau

**Comment by utkarshsaxena93** [comment]
Polaris modal is a good example to look at as a good example of some of the patterns around focus trapping and/or close buttons

File: assets/quick-order-list.js [hunk 2]
```diff
@@ -208,12 +212,38 @@ class QuickOrderList extends HTMLElement {
...
+      }
+    } else {
+      const stickyHeaderBottomBorder = this.stickyHeaderElement && this.stickyHeaderElement.getBoundingClientRect().bottom;
+      const totalBarCrossesInput = inputBottomBorder > this.totalBarPosition;
+      const inputOutsideOfViewPort = inputBottomBorder < this.inputFieldHeight;
+      const stickyHeaderCrossesInput = this.stickyHeaderElement && this.stickyHeader.type !== 'on-scroll-up' && this.stickyHeader.height > inputTopBorder;
+      const stickyHeaderScrollupCrossesInput = this.stickyHeaderElement && this.stickyHeader.type === 'on-scroll-up' && this.stickyHeader.height > inputTopBorder && stickyHeaderBottomBorder > 0;
+
+      if (totalBarCrossesInput || inputOutsideOfViewPort || stickyHeaderCrossesInput || stickyHeaderScrollupCrossesInput) {
+        this.variantListInput.scrollIntoView({ block: 'center', behavior: 'smooth' });
```
Comment (sofiamatulis) [inline assets/quick-order-list.js:234]: Just an idea but if this and what youre doing in line 224 are the same we can always make it a function to avoid duplicating the same line

And just call the function in the two places `this.scrollIntoView()`

```
scrollIntoView() {
 this.variantListInput.scrollIntoView({ block: 'center', behavior: 'smooth' });
}
```

Thoughts?

Comment (eugenekasimov) [reply assets/quick-order-list.js]: Love this idea even though there are just two repetitive lines, still looks better. Thanks for the idea @sofiamatulis

**Review by sofiamatulis — COMMENTED** [review]
Just to ensure I understand, we are not supposed to scroll back up correct?

https://screenshot.click/14-12-xmuox-88xps.mp4

Updated video: https://screenshot.click/14-32-lojam-j7bgz.mp4

**Comment by eugenekasimov** [comment]
> Just to ensure I understand, we are not supposed to scroll back up correct?
> 
> https://screenshot.click/14-12-xmuox-88xps.mp4
> 
> Updated video: https://screenshot.click/14-32-lojam-j7bgz.mp4

Nice catch @sofiamatulis . There was a bug and I fixed it [e1c902d](https://github.com/Shopify/dawn/pull/3343/commits/e1c902da1203447c4caa6f88d47d5f0ac5ef4552). Every time we change qty on any input for DOM we actually change the table and the table head accordingly. That's why after changing qty the calculation for table head stops working because we never get it again after updating. I used the same approach we did in other recent PR and added `getTableHead()`. 
Just for context, it has never been a case for other elements we used for position calculation because those other elements such as the Header or Total Bar never change while we update the table.

**Review by sofiamatulis — APPROVED** [review]
Thanks for fixing it 🔥 Let's just update the branch before merging https://screenshot.click/15-37-rh9x4-n0qsr.png
